src/TestProblem/Hydro/CCSN/Record_CCSN.cpp

```diff
@@ -451,3 +454,211 @@ void Detect_CoreBounce()
 
 } // FUNCTION : Detect_CoreBounce()
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Detect_Shock
+// Description :  Measure the minimum, maximum, and average shock radius if a shock is present
+//
+// Note        :  1. Invoked by Record_CCSN()
+//                2. The average shock radius is weighted by the inverse cell volume
+//                3. Ref: Balsara and Spicer, 1999, JCP, 149, 270, sec 2.2
+//-------------------------------------------------------------------------------------------------------
+void Detect_Shock()
+{
+
+#  ifdef OPENMP
+   const int NT = OMP_NTHREAD;
+#  else
+   const int NT = 1;
+#  endif
+
+   const int PS1P2   = PS1 + 2;
+   const int NPG_Max = FLU_GPU_NPGROUP;
+   const int Stride1 = CUBE( PS1P2 );
+   const int Stride2 = NCOMP_TOTAL * Stride1;
+   const int VELX    = NCOMP_PASSIVE + 1;
+   const int VELY    = NCOMP_PASSIVE + 2;
+   const int VELZ    = NCOMP_PASSIVE + 3;
+   const int PRES    = NCOMP_PASSIVE + 4;
+
+   const real   ThresholdFac_Pres = 0.5;
+   const real   ThresholdFac_Vel  = 0.1;
+   const double BoxCenter[3]      = { amr->BoxCenter[0], amr->BoxCenter[1], amr->BoxCenter[2] };
+
+
+   double Shock_Min    =  HUGE_NUMBER;
+   double Shock_Max    = -HUGE_NUMBER;
+   double Shock_Ave    =  0.0;
+   double Shock_Weight =  0.0;
+   bool   Shock_Found  =  false;
+
+   double OMP_Shock_Min   [NT];
+   double OMP_Shock_Max   [NT];
+   double OMP_Shock_Ave   [NT];
+   double OMP_Shock_Weight[NT];
+   bool   OMP_Shock_Found [NT];
+
+   real *OMP_Fluid    = new real [ 8*NPG_Max*Stride2 ];
+   real *OMP_Pres_Min = new real [ 8*NPG_Max*Stride1 ];
+   real *OMP_Cs_Min   = new real [ 8*NPG_Max*Stride1 ];
+
+   for (int t=0; t<NT; t++)
+   {
+      OMP_Shock_Min   [t] =  HUGE_NUMBER;
+      OMP_Shock_Max   [t] = -HUGE_NUMBER;
+      OMP_Shock_Ave   [t] = 0.0;
+      OMP_Shock_Weight[t] = 0.0;
+      OMP_Shock_Found [t] = false;
+   }
+
+
+   for (int lv=0; lv<NLEVEL; lv++)
+   {
+      const double dh        = amr->dh[lv];
+      const double _dv       = 1.0 / CUBE( dh );
+      const int    NTotal    = amr->NPatchComma[lv][1] / 8;
+            int   *PID0_List = new int [NTotal];
+
+      for (int t=0; t<NTotal; t++)  PID0_List[t] = 8*t;
+
+      for (int Disp=0; Disp<NTotal; Disp+=NPG_Max)
+      {
+         int NPG = ( NPG_Max < NTotal-Disp ) ? NPG_Max : NTotal-Disp;
+
+//       (1-a) prepare the primitive and passive variables
+//             note that the data are prepared in order of density, Passive, velx, vely, velz, and pressure
+         Prepare_PatchData( lv, Time[lv], OMP_Fluid, NULL,
+                            1, NPG, PID0_List+Disp, _DENS|_PASSIVE|_VELX|_VELY|_VELZ|_PRES, _NONE,
+                            OPT__FLU_INT_SCHEME, INT_NONE, UNIT_PATCH, NSIDE_26, false,
+                            OPT__BC_FLU, BC_POT_NONE, -1.0, -1.0, -1.0, -1.0, false );
+
+
+#        pragma omp parallel for schedule( runtime )
```

Comment:
just curious, is this function done in CPU only? or can be used in GPU as well?

Reply:
This function uses CPU only as others for recording function (e.g., GW). I think it is possible to use GPU, but the code will be more complicated.